torch/distributed/pipelining/_schedule_visualizer.py

```diff
@@ -130,33 +131,71 @@ def visualize_schedule(
         draw_position = 0  # Initialize drawing position for each rank
         for action in actions:
             if action is not None:
-                comp_type_color = action_type_to_color_mapping.get(
-                    action.computation_type, _ComputationTypeColor("black")
-                )
-                used_computation.add(action.computation_type)
-                color = comp_type_color.color
-                width = comp_type_color.width
-                # Draw the rectangle to represent the action duration
-                rect = Rectangle(
-                    (draw_position, num_ranks - rank_idx - 1),
-                    width,
-                    1,
-                    facecolor=color,
-                    edgecolor="black",
-                )
-                ax.add_patch(rect)
-                # Draw the text centered within the rectangle
-                ax.text(
-                    draw_position + width / 2,
-                    num_ranks - rank_idx - 1 + 0.5,
-                    str(action),
-                    ha="center",
-                    va="center",
-                    fontsize=font_size,
-                    color="white",
-                )
-                # Increment the drawing position by the width of the current action
-                draw_position += width
+                # Check if action has sub_actions
+                if action.sub_actions is not None:
```

Comment:
seems like you could avoid copypaste most of the if/else branches.  only a few details differ.  can you pull `comp_type_color`, `used_computation.add`, `color=`, `width=` up above the if/else, and put rect = and ax.text below the if/else?  you just need to determine a few things inside the if/else like the linewidth, text properties etc.